wire-compiler/src/main/java/com/squareup/wire/WireCompiler.java

```diff
@@ -225,39 +232,73 @@ void compile() throws IOException {
         .withAndroid(emitAndroid)
         .withCompact(emitCompact);
 
+    ExecutorService executor = Executors.newCachedThreadPool();
+    ConcurrentLinkedQueue<Type> types = new ConcurrentLinkedQueue<>();
     for (ProtoFile protoFile : schema.protoFiles()) {
       // Check if we're skipping files not explicitly named.
       if (!sourceFileNames.isEmpty() && !sourceFileNames.contains(protoFile.location().path())) {
         if (namedFilesOnly || protoFile.location().path().equals(DESCRIPTOR_PROTO)) continue;
       }
+      types.addAll(protoFile.types());
+    }
 
-      for (Type type : protoFile.types()) {
-        TypeSpec typeSpec = javaGenerator.generateType(type);
-        ClassName javaTypeName = (ClassName) javaGenerator.typeName(type.type());
-        writeJavaFile(javaTypeName, typeSpec, type.location());
+    List<Future<Void>> futures = new ArrayList<>(8);
+    for (int i = 0; i < MAX_WRITE_CONCURRENCY; ++i) {
+      futures.add(i, executor.submit(new JavaFileWriter(javaGenerator, types)));
+    }
+    executor.shutdown();
+
+    try {
+      for (Future<Void> future : futures) {
+        future.get();
```

Comment:
Got it. Futures are such boilerplate-ridden messes in pre-Java 8... we have Guava and I was looking into what it affords but perhaps we just want to enforce JDK 8 in the compiler.

Guava has a couple helpers on `Futures` that might make this simpler, but it also requires the use of `ListenableFuture`.

With Java 8 this looks like it could be done trivially with `.parallelStream()` , although I'm less certain how that deals with exceptions. There's also `CompleteableFuture` in that version which is like `ListenableFuture`++.